lib/fuse_lowlevel.c

```diff
@@ -320,38 +320,46 @@ int fuse_reply_iov(fuse_req_t req, const struct iovec *iov, int count)
 	return res;
 }
 
-static size_t fuse_dirent_size(size_t namelen)
-{
-	return FUSE_DIRENT_ALIGN(FUSE_NAME_OFFSET + namelen);
-}
+#ifndef IFTODT
+# define IFTODT(mode) (((mode) & 0170000) >> 12)
+#endif
 
-static void fuse_add_dirent(struct fuse_dirent *dirent, const char *name,
-			    const struct stat *stbuf, off_t off)
+static void fuse_add_dirent(struct fuse_dirent *dirent,
+                            const char *name, const size_t namelen,
+                            const size_t entlen, const size_t entlen_padded,
+			    const struct stat *stbuf, const off_t off)
 {
-	unsigned namelen = strlen(name);
-	unsigned entlen = FUSE_NAME_OFFSET + namelen;
-	unsigned entsize = fuse_dirent_size(namelen);
-	unsigned padlen = entsize - entlen;
+	unsigned padlen = entlen_padded - entlen;
 
 	dirent->ino = stbuf->st_ino;
 	dirent->off = off;
 	dirent->namelen = namelen;
-	dirent->type = (stbuf->st_mode & 0170000) >> 12;
+	dirent->type = IFTODT(stbuf->st_mode);
 	strncpy(dirent->name, name, namelen);
 	if (padlen)
 		memset(dirent->name + namelen, 0, padlen);
```

Comment:
Yes, but why do the values need to get passed to `fuse_add_dirent`? But you are right about the clearining be a little clumy because of the need to cast.